### PR TITLE
feat(search): search 2.0

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR revamps Semgrep's IDE search to have a more accessible UI, as well as being more responsive and more expressive in some ways.

## Why:
Semgrep's search is currently nice, but I wouldn't use it because of several UX troubles. Currently, you have to go through a bunch of buttons clicks, and search is pretty slow to boot.

## How:
We deprecated the VS Code API based UI for one which is entirely in React, in the form of a webview. This lets us have full customizability over what the UI looks like.

We also made several engine improvements which are consumed by the new UI to offer better UX, such as:
- the ability to write multiple patterns (including negated patterns) to refine search
- semantic fixes
- streaming searches
- includes/excludes based on paths

## Test plan:
Thorough manual testing.

image

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [ ] A changelog entry was for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

**Comment by brandonspark** [comment]
(tagging @bkettle if you have any concerns)

File: src/views/webview.ts [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+  // Send messages from extension -> webview
+  sendMessageToWebview(data: extensionToWebviewCommand): void {
+    this._view?.webview.postMessage(data);
+  }
+
+  public resolveWebviewView(
+    webviewView: vscode.WebviewView
+    // these are unused, so commented out
+    // context: vscode.WebviewViewResolveContext,
+    // _token: vscode.CancellationToken
```
Comment (ajbt200128) [inline src/views/webview.ts:126]: any reason to leave these comments in?

File: src/views/webview.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      vscode.Uri.joinPath(assetsPath, "webview.js")
+    );
+    // The global CSS file
+    const globalStylesUri = webview.asWebviewUri(
+      vscode.Uri.joinPath(this._extensionUri, "media", "global.css")
+    );
+
+    const nonce = randomUUID();
+
+    // Tip: Install the es6-string-html VS Code extension to enable code highlighting below
```
Comment (ajbt200128) [inline src/views/webview.ts:163]: you can add this to https://github.com/semgrep/semgrep-vscode/blob/master/.vscode/extensions.json so people get prompted to install it

File: src/views/webview.ts [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+
+  //       switch (command) {
+  //         case "hello":
+  //           vscode.window.showInformationMessage(text);
+  //           return;
+  //       }
+  //     },
+  //     undefined,
+  //     this._disposables
+  //   );
```
Comment (ajbt200128) [inline src/views/webview.ts:199]: delete?

File: src/webview-ui/src/components/SearchResults/MatchItem.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      <>
+        <span className={styles.matchTextDeleted}>{inside}</span>
+        <span className={styles.matchTextInserted}>
+          {match.searchMatch.fix}
+        </span>
+      </>
+    );
+  } else {
+    matchText = <span className={styles.matchTextNormal}>{inside}</span>;
+  }
```
Comment (ajbt200128) [inline src/webview-ui/src/components/SearchResults/MatchItem.tsx:34]: v small nit: People usually just would use the ternary operator here

Comment (ajbt200128) [reply src/webview-ui/src/components/SearchResults/MatchItem.tsx:34]: just a common style when choosing between html blocks in react for whatever reason

File: src/webview-ui/src/components/SearchResults/SearchResultEntry.tsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+export interface SearchResultEntryProps {
+  result: ViewResult;
+}
+export const SearchResultEntry: React.FC<SearchResultEntryProps> = ({
+  result,
+}) => {
+  const [isExpanded, setIsExpanded] = useState(true);
+  // This is really only so that we can re-render.
+  const [numRerenders, setNumRerenders] = useState(0);
```
Comment (ajbt200128) [inline src/webview-ui/src/components/SearchResults/SearchResultEntry.tsx:17]: this is an antipattern. Abusing useState and/or `forceUpdate` is something you should avoid. I'd try and figure out what state is changing that makes you want to rerender, and use state there. It looks like it's the matches, so I'd probably do a useState for that. Even tho it doesn't really do anything cause pointers, it still makes the code easier to understand, esp for the avg react developer

File: src/webview-ui/src/components/SearchResults/SearchResultEntry.tsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+  function onFix(match: ViewMatch) {
```
Comment (ajbt200128) [inline src/webview-ui/src/components/SearchResults/SearchResultEntry.tsx:19]: Usually the style here is to just define an anonymous function. That way if you want to redefine them for whatever reason you can.

File: src/webview-ui/src/hooks/useStore.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    ],
+    fix: fixValue,
+    includes: includes,
+    excludes: excludes,
+    scanID: scanID,
+    lang,
+  });
+}
+
+export function useStore(key: keyof Store): [any, (value: any) => void] {
```
Comment (ajbt200128) [inline src/webview-ui/src/hooks/useStore.ts:73]: not sure I understand why we've implemented useStore here. Usually w/react you have a tree of components passing info down, i.e. pattern/includes/excludes, that way if whatever the source is changes then all components down the tree rerender based on the new props. The components should be idempotent based on the props. So here App.tsx would pass whatever youve been using useStore for down to it's children that need it. Same w.r.t. to the setters. This also simplifies your code as then you really understand what components mutate global state, and which don't need to. And then you don't have unused state setters/getters  anywhere. And then you also get type checking too. 

TLDR; Lift state up, don't use useStore, this explains it better than I do: https://react.dev/learn/sharing-state-between-components

I'm not sure the lifetime of App.tsx, but I saw the webview option that is like "keep context" or something before, so I suspect you may be fine without using local storage. If not that's also ok, but I'd just do a useState or something and not have a dedicated useStore.

Comment (ajbt200128) [reply src/webview-ui/src/hooks/useStore.ts:73]: the whole section of "Managing state" may help clean up this code some more if you want, but idc too much beyond the useStore since I don't think this webview will get much more complicated, but if it does it may be worth reading through the ones about context

File: src/webview-ui/src/components/TopSection/PatternBadge.tsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+export const PatternBadge: React.FC<PatternBadgeProps> = ({
+  onNewPattern,
+  index,
+  patterns,
+  positive,
+  onPositivityToggle,
+}) => {
+  const last = isLast(index, patterns);
+  const color = positive ? "#458c4c" : "#a23636";
+  const heightOfAdd = last ? "15px" : "27px";
```
Comment (ajbt200128) [inline src/webview-ui/src/components/TopSection/PatternBadge.tsx:22]: throw these in a css file plz lmao

File: src/webview-ui/src/components/utils/LangChooser.tsx [hunk 9]
```diff
@@ -0,0 +1,69 @@
+import { VSCodeDropdown, VSCodeOption } from "@vscode/webview-ui-toolkit/react";
+import { vscode } from "../../../utilities/vscode";
+import { SearchLanguage } from "../../../../interface/interface";
+import { useState } from "react";
+import { Store, useStore } from "../../hooks/useStore";
+import { SUPPORTED_LANGS } from "../../../../constants";
+
+const style = {
```
Comment (ajbt200128) [inline src/webview-ui/src/components/utils/LangChooser.tsx:8]: css file

File: src/webview-ui/src/hooks/useStore.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  };
+}
+
+const store: Store = defaultStore();
+
+export function generateUniqueID(): string {
+  return Math.random().toString(36).substring(7);
+}
+
+export function useSearch(onNewSearch: (scanID: string) => void): void {
```
Comment (ajbt200128) [inline src/webview-ui/src/hooks/useStore.ts:43]: this is fine because it's not setting any global state anywhere (I think)

Comment (ajbt200128) [reply src/webview-ui/src/hooks/useStore.ts:43]: but should probably take all the store values as a param, not directly access them

File: src/webview-ui/utilities/vscode.ts [hunk 11]
```diff
@@ -28,14 +55,56 @@ class VSCodeAPIWrapper {
...
+      }
+      case "extension/semgrep/exportRuleRequest": {
+        if (this.onExportRule) {
+          this.onExportRule();
+        }
+      }
     }
   }
 
+  /* We shouldn't actually need this code for now, because we are storing the
```
Comment (ajbt200128) [inline src/webview-ui/utilities/vscode.ts:104]: lets get rid of it if we don't need it

**Review by ajbt200128 — APPROVED** [review]
A few nits w.r.t. style w/functionsm css and stuff. Big thing is how we handle state here isn't very react-y. Approving to unblock you but would like to see the state stuff refactored (and retested) before merging

File: src/env.ts [hunk 12]
```diff
@@ -33,10 +33,19 @@ export class Environment {
...
+  private _provider: SemgrepSearchWebviewProvider | null = null;
+  /* The scan ID is the (hopefully) unique identifier associated to each
+     /semgrep/search request.
+     The reason why we need it is for synchronization, in the event that
+     a user issues a scan while another one is still completing. We don't
+     have a good way of reaching out to each individual searchLoop()
+     (which is asynchronous) and telling it to stop, so we change this
+     mutable variable so that it knows to stop on its own.
+   */
+  private _scanID: string | null = null;
```
[the remark this answers is not in the document]
Comment (brandonspark) [reply src/env.ts:45]: im pretty sure i killed the search provider, so i think that this needs to go in the environment so that the `commands.ts` hook can share the information

File: src/views/webview.ts (hunk 1 above)
Comment (brandonspark) [reply src/views/webview.ts:126]: i was having some difficulties where eslint would keep yelling at me even if I prefixed with `_`, but i fixed those so this isn't necessary anymore

**Comment by brandonspark** [comment]
@ajbt200128 let me know how you feel about the current state management. i tried to pull all the state to the top of the `App` via a new `useStore` which simply gives the entire `store` object, which is then modified via a `useSetStore` hook which causes rerendering of the tree

File: package.json [hunk 13]
```diff
@@ -87,9 +87,9 @@
         "when": "semgrep.cli.minor >= 25 || config.semgrep.ignoreCliVersion"
       },
       {
-        "command": "semgrep.search.replaceAll",
-        "title": "Semgrep Search: Replace All Matches",
-        "icon": "$(tools)",
+        "command": "semgrep.search.exportRule",
+        "title": "Semgrep Search: Export Rule",
+        "icon": "$(search-remove)",
```
[the remark this answers is not in the document]
Comment (brandonspark) [reply package.json:86]: how does documentation typically work in this repo? should I add it directly to the `CHANGELOG.md`, or is there a programmatic interface for this?

File: yarn.lock
Comment (ajbt200128) [inline yarn.lock:1]: delete
